p2p/src/authenticated/lookup/actors/tracker/actor.rs

```diff
@@ -0,0 +1,427 @@
+        // Sign my own information
+        let myself = (cfg.crypto.public_key(), cfg.address);
+
+        // General initialization
+        let directory_cfg = directory::Config {
+            mailbox_size: cfg.mailbox_size,
+            max_sets: cfg.tracked_peer_sets,
+            rate_limit: cfg.allowed_connection_rate_per_peer,
+            allow_private_ips: cfg.allow_private_ips,
+        };
+        let directory = Directory::init(context.clone(), myself, directory_cfg);
+        let (sender, receiver) = mpsc::channel(cfg.mailbox_size);
+
+        (
+            Self {
+                context,
+                max_peer_set_size: cfg.max_peer_set_size,
+                receiver,
+                directory,
+            },
+            Mailbox::new(sender.clone()),
+            Oracle::new(sender),
+        )
+    }
+
+    /// Start the actor and run it in the background.
+    pub fn start(mut self) -> Handle<()> {
+        self.context.spawn_ref()(self.run())
+    }
+
+    async fn run(mut self) {
+        while let Some(msg) = self.receiver.next().await {
+            self.directory.process_releases();
+            match msg {
+                Message::Register { index, peers } => {
+                    // Ensure that peer set is not too large.
+                    // Panic since there is no way to recover from this.
+                    let len = peers.len();
+                    let max = self.max_peer_set_size;
+                    assert!(len <= max, "peer set too large: {} > {}", len, max);
+
+                    self.directory.add_set(index, peers);
+                }
+                Message::Connect {
+                    public_key,
+                    mut peer,
+                } => {
+                    // Kill if peer is not authorized
+                    if !self.directory.allowed(&public_key) {
+                        peer.kill().await;
+                        continue;
+                    }
+
+                    // Mark the record as connected
+                    self.directory.connect(&public_key);
+                }
+                Message::Dialable { responder } => {
+                    let _ = responder.send(self.directory.dialable());
+                }
+                Message::Dial {
+                    public_key,
+                    reservation,
+                } => {
+                    let _ = reservation.send(self.directory.dial(&public_key));
+                }
+                Message::Listen {
+                    public_key,
+                    reservation,
+                } => {
+                    let _ = reservation.send(self.directory.listen(&public_key));
+                }
+                Message::Block { public_key } => {
+                    // Block the peer
+                    self.directory.block(&public_key);
+
+                    // We don't have to kill the peer now. It will be sent a `Kill` message the next
```

Comment:
We now do need to kill the peer explicitly (as we only send `Connect` at the start). We should call `Kill` both in `Register` if no longer allowed and in `Block`.

Alternatively, we can restore the now dropped `BitVec` as a `Heartbeat` that can trigger this `Kill`.